### PR TITLE
Refactor queue rspec runner

### PR DESCRIPTION
**Pull request by ArturT**
# story
https://trello.com/c/BcBhX2MW

# related

* https://github.com/KnapsackPro/knapsack_pro-ruby/pull/237

# How to mock the Queue API responses in the knapsack_pro gem

You can add the following code to the source code of the Knapsack Pro gem to mock the Queue API responses.
https://github.com/KnapsackPro/knapsack_pro-ruby/pull/237/commits/348f563a1cc5d72ac1684aa3848aa7f749f9a738

```ruby
# lib/knapsack_pro/queue_allocator.rb

    def test_file_paths(can_initialize_queue, executed_test_files)
      if ENV['MOCK_QUEUE_API_RESPONSE'] == 'true'
        @@index ||= 0
        batches = [
          ['spec/features/calculator_spec.rb[1:2:1]', 'spec/controllers/articles_controller_spec.rb'],
          ['spec/collection_spec.rb'],
          ['spec/features/calculator_spec.rb[1:1:1]'],
          ['spec/bar_spec.rb'],
          [], # the last Queue API response is always an empty list of test files
        ]

        # example of an empty batch for a late CI node
        #batches = [
          #[]
        #]
        tests = batches[@@index] || []
        @@index += 1
        puts '='*50
        puts 'Tests (mocked API response):'
        puts tests.inspect
        puts '='*50
        return tests
      end

      return [] if @fallback_activated
...
```

### CODE REVIEW

File: bin/edge_cases/knapsack_pro_queue_rspec_record_first_run [hunk 1]
```diff
@@ -27,6 +27,13 @@ BRANCH_NAME=fake-branch
 #export KNAPSACK_PRO_RSPEC_SPLIT_BY_TEST_EXAMPLES=true
 #export KNAPSACK_PRO_SLOW_TEST_FILE_PATTERN="spec/**{,/*/**}/*_spec.rb"
 
+# In order to mock the Queue API responses, you need to change source code of the knapsack_pro gem:
```
Comment (ArturT) [inline bin/edge_cases/knapsack_pro_queue_rspec_record_first_run:30]: The purpose of this file was to help developer (me) to play with the knapsack_pro gem and a Rails app at the same time, so that I could see what will happen on the output, I could use a debugger etc.

We have isolated integration tests in PR https://github.com/KnapsackPro/knapsack_pro-ruby/pull/237
The isolated integration tests are already covering all edge cases (except real API calls). 

This bin script is in `bin/edge_cases` directory. It is not executed as part of `bin/knapsack_pro_all` bin script ([see](https://github.com/KnapsackPro/rails-app-with-knapsack_pro/blob/master/bin/knapsack_pro_all.rb)).

File: spec/support/config/rspec_retry.rb [hunk 2]
```diff
@@ -5,9 +5,9 @@
   config.verbose_retry = true
   # show exception that triggers a retry if verbose_retry is set to true
   config.display_try_failure_messages = true
-
-  # It's the RSpec option. When too many tests fail then the Knapsack Pro is terminated
-  # config.fail_fast = 3
+ 
+  # --fail-fast option makes RSpec fail after X failed tests. This can lead to canceled tests in Queue Mode.
+  #config.fail_fast = 1
```
Comment (3v0k4) [inline spec/support/config/rspec_retry.rb]:
```suggestion
  # config.fail_fast = 1
```

**Review by 3v0k4 — APPROVED** [review]
I don't have much to say about this, but I hope that going forward we'll wipe this repo 🙂

**Comment by ArturT** [comment]
> I don't have much to say about this, but I hope that going forward we'll wipe this repo 🙂

We are moving in this direction but I think we still need it so that we could easily test a Rails app + Knapsack Pro integration on various CI providers.
